pkg/services/store/entity/sqlstash/watch_pg_notify_listen.go

```diff
@@ -0,0 +1,125 @@
+package sqlstash
+
+import (
+	"context"
+	"database/sql/driver"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/lib/pq"
+
+	"github.com/grafana/grafana/pkg/services/store/entity"
+)
+
+type pgNotifyPayload struct {
+	ResourceVersion string `json:"resource_version"`
+	Key             string `json:"key"`
+}
+
+// pgWatcher is like poller but uses the native postgres LISTEN/NOTIFY SQL commands
+func (s *sqlEntityServer) pgWatcher(stream chan *entity.Entity) {
+	ctx := context.Background()
+
+	engine, err := s.db.GetEngine()
+	if err != nil {
+		s.log.Error("error getting engine: %w", err)
+		return
+	}
+
+	conn, err := engine.DB().DB.Conn(ctx)
+	if err != nil {
+		s.log.Error("error getting db connection: %w", err)
+		return
+	}
+
+	defer func() { _ = conn.Close() }()
+
+	err = conn.Raw(func(driverConn any) error {
+		dc, ok := driverConn.(driver.Conn)
+		if !ok {
+			return fmt.Errorf("cannot convert driverConn from %T to driver.Conn", driverConn)
+		}
+
+		pq.SetNotificationHandler(dc, func(n *pq.Notification) {
+			if n == nil {
+				s.log.Debug("received postgres notification", "nil", true)
+				return
+			}
+			s.log.Debug("received postgres notification", "channel", n.Channel, "payload", n.Extra)
+
+			go func() {
+				var ne pgNotifyPayload
+				if err := json.NewDecoder(strings.NewReader(n.Extra)).Decode(&ne); err != nil {
+					s.log.Error("error decoding postgres notification with JSON payload %q: %w", n.Extra, err)
+					return
+				}
+
+				rv, err := strconv.ParseInt(ne.ResourceVersion, 10, 64)
+				if err != nil {
+					s.log.Error("error parsing resource version from postgres notification payload %q: %w", n.Extra, err)
+					return
+				}
+
+				r, err := s.Read(ctx, &entity.ReadEntityRequest{
+					Key:             ne.Key,
+					ResourceVersion: rv,
+					WithBody:        true,
+					WithStatus:      true,
+				})
+				if err != nil {
+					s.log.Error("error reading entity for postgres notification payload %q: %w", n.Extra, err)
+					return
+				}
+
+				stream <- r
+			}()
+		})
+
+		return nil
+	})
+	if err != nil {
+		s.log.Error("error creating new postgres connection: %w", err)
+		return
+	}
+
+	_, err = conn.ExecContext(ctx, `
+create or replace function tgf_notify_entity_history() returns trigger as $BODY$
+	begin
+		perform pg_notify('unifiedstorage', json_build_object('key', new.key::text, 'resource_version', new.resource_version::text)::text);
+		return new;
+	end
+$BODY$ language plpgsql;
+
+create or replace trigger tg_notify_entity_history after insert on entity_history for each row execute function tgf_notify_entity_history();
+
+LISTEN unifiedstorage;`)
```

Comment:
I suggest we use the channel name as a way of segmenting notifications. My original proposal was using the k8s namespace (a.k.a. the tenant) so we can also isolate their notifications and they don't interfere. But I see postgres has a limit of 63 bytes on channel names (and all identifiers). In my local, I see my user has namespace `default`. Do you think we can use 63 bytes for the tenant somehow? If the namespace becomes too large like an FQDN-like identifier we may fall short.

Reply:
I'm unsure how much we gain by doing that, especially as most internal operators are going to end up watching across namespaces. It would certainly mean that in a multitenant environment we end up with a whole lot of per-tenant overhead that we don't really need when we can run things through a single channel.

Reply:
Interesting. I was working under the assumption that a single user (even service accounts) had to be constrained to a single namespace, namely, their orgID.